### DISCUSSION TITLE
What is the recommended project structure strategy?

### DISCUSSION THREAD

**Discussion by viktorKhan** [lucoenergia/conluz-web#1] [Ideas]
Currently, the project structure is as follow:
![imagen](https://github.com/lucoenergia/conluz-web/assets/4686378/6f1e61f9-7a04-4686-874e-905ce5939509)

But this seems not be the best approach. 

I specially see not good to having the word "modules" in the navigator path every time I navigate to a page. For instance:

- `http://localhost:3000/modules/dashboard`
- `http://localhost:3000/modules/login`

So, what is the recommended way to structure the project?

**Comment by viktorKhan** [comment]
Asking ChatGPT 3.5, this is its answer:

The recommended project structure for a Next.js project can vary based on personal preferences and the specific requirements of your project. However, there are some commonly accepted practices that you can follow to organize your Next.js project effectively. Below is a general project structure that you might find useful:

```plaintext
my-nextjs-app/
|-- pages/
|   |-- index.js
|   |-- about.js
|   |-- contact.js
|-- components/
|   |-- Header.js
|   |-- Footer.js
|   |-- Button.js
|-- styles/
|   |-- globals.css
|   |-- Home.module.css
|   |-- About.module.css
|   |-- Contact.module.css
|-- public/
|   |-- images/
|       |-- logo.png
|   |-- favicon.ico
|-- utils/
|   |-- api.js
|   |-- helperFunctions.js
|-- .gitignore
|-- package.json
|-- README.md
|-- next.config.js
```

Explanation of the structure:

- **pages**: This directory is special in Next.js, as each `.js` or `.jsx` file inside it becomes a route in your application. For example, `pages/index.js` corresponds to the root URL ("/"), and `pages/about.js` corresponds to "/about". Nested folders also create nested routes.

- **components**: This directory is for your React components that are not pages. These can be reusable pieces of UI that you use across multiple pages.

- **styles**: This directory is for your styles. You can use plain CSS, CSS modules, or other styling solutions here.

- **public**: This directory is used to serve static assets like images. Files inside the public directory can be referenced in your code using the base URL (`/`). For example, if you have an image in `public/images/logo.png`, you can use it in your code like `<img src="/images/logo.png" />`.

- **utils**: This directory is for utility functions and modules that you might use across your application.

- **.gitignore**: Specifies files that should be ignored by Git. It's a good practice to include files like `node_modules` and build artifacts.

- **package.json**: Contains metadata about your project, as well as the list of dependencies. You can manage scripts, dependencies, and other project-related settings here.

- **README.md**: Documentation for your project.

- **next.config.js**: Configuration file for Next.js. You can use it to customize various aspects of the Next.js build process.

Keep in mind that this is just a suggested structure, and you can modify it based on your specific needs. Additionally, as your project grows, you may consider further modularization and organization based on features or domain-driven design principles.

**Reply by ACD03** [reply]
An approximation of ChatGPT's suggestion is the project structure detailed on this [site](https://makerkit.dev/docs/next-fire/application-structure). Essentially, the site recommends using the following main folders:
- **"lib":**  contains the app's business logic components, such as hooks, context, and queries.
- **"components":** contains the view code for components.
- **"pages":** contains the final views for each app route.

Although this suggestion could be a starting point, I believe that it needs to be adapted to Next.js App Routing because, like the suggestion from ChatGPT, the site's recommendation is based on Next.js Page Routing.

**Comment by viktorKhan** [comment]
Here is the recommendation from the official web page -> https://nextjs.org/docs/getting-started/project-structure

According to that, this is the recommended structure for top-level forlders:
![imagen](https://github.com/lucoenergia/conluz-web/assets/4686378/df5f3a2b-6413-4e53-a67d-ae5297de4965)

**Comment by ACD03** [comment]
The first approach is an adaptation of the [React Native suggestion](https://www.reactnative.express/app/project_structure). This site provides details on how React projects could be organized based on the project's size. However, it also recommends using the presentational/container pattern, which, in my opinion and based in other [articles](https://www.freecodecamp.org/news/separation-of-concerns-react-container-and-presentational-components/), could be substituted by the newest custom hooks feature.

I agree with the idea that the word "modules" should not appear in the path of each page. Additionally, I believe that this project structure is not the most suitable for this case because the author of the React Native site does not use the Next.js framework and its routing mode.

**Comment by ACD03** [comment]
Another suggestion can be found on Materio´s [site](https://demos.themeselection.com/materio-mui-react-nextjs-admin-template/documentation/docs/guide/development/folder-structure), which is the template used in this project. In my opinion, this structure is more complex than the aforementioned suggestions, although it is true that it could also be an option to consider.

**Comment by viktorKhan** [comment]
Finally, we have decided to go with the following structure:

```
├── LICENSE
├── next.config.js
├── next-env.d.ts
├── node_modules
├── package.json
├── package-lock.json
├── postcss.config.js
├── public
├── README.md
├── src
│	├── account-settings
│	├── AppLayout.tsx
│	├── cards
│	├── dashboard
│	├── error
│	├── form-layouts
│	├── icons
│	├── layout.tsx
│	├── login
│	├── NavigationEvents.tsx
│	├── page.tsx
│	├── register
│	├── shared
│	│	├── components
│	│	├── configs
│	│	├── context
│	│	├── css
│	│	├── fonts
│	│	├── hooks
│	│	├── layouts
│	│	├── styles
│	│	└── utils
│	├── tables
│	└── typography
├── tailwind.config.js
└── tsconfig.json
```
Where:
- Each route of the app will have a folder (and sub-folders with more routes if apply) within `src/app`
  - Within each of these folders, we could have sub-folders to store elements that must be separated from the representational part like: hooks or components.
- All the common parts that could be used in different parts of the application will be included in the folder `src/app/shared`
- `public`  folder will store elements that must be publicly accessible like images.
